pallets/spin-polkadot/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+            let mut weight = Weight::zero();
+            weight += T::DbWeight::get().reads(1);
+
+            match <State<T>>::get() {
+                SlowchainState::Operational {
+                    last_alive_message_block_number,
+                } => {
+                    let timeout_blocks: BlockNumberFor<T> = T::TimeoutBlocks::get().into();
+
+                    let deadline_block_number: Option<BlockNumberFor<T>> =
+                        last_alive_message_block_number.checked_add(&timeout_blocks);
+                    let deadline_block_number = match deadline_block_number {
+                        Some(deadline_block_number) => deadline_block_number,
+                        None => return weight,
+                    };
+
+                    if current_block_number > deadline_block_number {
+                        log::info!(
+                            "on_initialize: alive message deadline exceeded. Starting cool down"
+                        );
+                        <State<T>>::put(SlowchainState::CoolDown {
+                            start_block_number: current_block_number,
+                        });
+                        weight += T::DbWeight::get().writes(1);
+                        Self::deposit_event(Event::StartedCoolDown {
+                            block_number: current_block_number,
+                            last_alive_message_block_number,
+                        });
+                    }
+                }
+                SlowchainState::CoolDown { start_block_number } => {
+                    let cool_down_period_blocks: BlockNumberFor<T> =
+                        T::CoolDownPeriodBlocks::get().into();
+                    let cool_down_period_deadline: Option<BlockNumberFor<T>> =
+                        start_block_number.checked_add(&cool_down_period_blocks);
+                    let cool_down_period_deadline = match cool_down_period_deadline {
+                        Some(cool_down_period_deadline) => cool_down_period_deadline,
+                        None => return weight,
+                    };
+                    if current_block_number > cool_down_period_deadline {
+                        <State<T>>::put(SlowchainState::Operational {
+                            last_alive_message_block_number: current_block_number,
+                        });
+                        weight += T::DbWeight::get().writes(1);
+                        Self::deposit_event(Event::FinishedCoolDown {
+                            block_number: current_block_number,
+                        });
+                    }
+                }
+            }
+
+            weight
+        }
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::call_index(0)]
+        #[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().reads_writes(1,1))]
+        pub fn handle_alive_message(
+            origin: OriginFor<T>,
+            proof: AliveMessageProof<HeaderFor<T>>,
```

Comment:
I'm wondering if this works for the other chain `HeaderFor<T>` :thinking:

Did you test it or let's see during an integration to the runtime?

Reply:
Let's test and we'll see 🤔